src/libraries/System.Private.Uri/src/System/IriHelper.cs

```diff
@@ -263,18 +260,16 @@ internal static unsafe string EscapeUnescapeIri(char* pInput, int start, int end
                 {
                     const int MaxNumberOfBytesEncoded = 4;
 
-                    byte[] encodedBytes = new byte[MaxNumberOfBytesEncoded];
-                    fixed (byte* pEncodedBytes = &encodedBytes[0])
-                    {
-                        int encodedBytesCount = Encoding.UTF8.GetBytes(pInput + next, surrogatePair ? 2 : 1, pEncodedBytes, MaxNumberOfBytesEncoded);
-                        Debug.Assert(encodedBytesCount <= MaxNumberOfBytesEncoded, "UTF8 encoder should not exceed specified byteCount");
+                    Debug.Assert(sizeof(int) == MaxNumberOfBytesEncoded);
+                    int encodedBytesBuffer;
+                    byte* pEncodedBytes = (byte*)&encodedBytesBuffer;
 
-                        bufferRemaining -= encodedBytesCount * percentEncodingLen;
+                    int encodedBytesCount = Encoding.UTF8.GetBytes(pInput + next, surrogatePair ? 2 : 1, pEncodedBytes, MaxNumberOfBytesEncoded);
+                    Debug.Assert(encodedBytesCount <= MaxNumberOfBytesEncoded, "UTF8 encoder should not exceed specified byteCount");
 
-                        for (int count = 0; count < encodedBytesCount; ++count)
-                        {
-                            UriHelper.EscapeAsciiChar((char)encodedBytes[count], ref dest);
-                        }
+                    for (int count = 0; count < encodedBytesCount; ++count)
+                    {
+                        UriHelper.EscapeAsciiChar((char)*(pEncodedBytes + count), ref dest);
```

Comment:
Can we use this as an opportunity to get rid of some unsafe code and just use spans? e.g.
```C#
Span<byte> encodedBytes = stackalloc byte[MaxNumberOfBytesEncoded];
int encodedBytesCount = Encoding.UTF8.GetBytes(new ReadOnlySpan<byte>(pInput + next, surrogatePair ? 2 : 1), encodedBytes);
for (int count = 0; count < encodedBytesCount; count++)
{
    UriHelper.EscapeAsciiChar((char)encodedBytes[i], ref dest);
}
```

Reply:
I see a ~2% perf hit on the benchmark by doing so.
If we're okay with that I can make the change.

Reply:
2% might be a measurement noise. I just ran in the same issue last week where the perf deviated by +- 2% between runs without any code changes.

Reply:
What is the microbenchmark?

Reply:
Try running perf test multiple time and measure deviation. Additionally, you can set CPU affinity for the benchmark process it can stabilize results a bit.

Reply:
> What is the microbenchmark?

Just calling the Uri ctor with `"scheme:" + { '\ud83f', '\udffe' } * 1000` as input.

I can try isolating the above code to a more-micro benchmark.

Reply:
In a [microbenchmark](https://gist.github.com/MihaZupan/628bf37fe3d815183bb3d93fba337afc), the difference is ~25%.

Reply:
I think @stephentoub's sample is probably doing a bounds check for each byte. If so, it's trivial to optimize that out:

```c#
Span<byte> encodedBytes = stackalloc byte[MaxNumberOfBytesEncoded];

int encodedBytesCount = Encoding.UTF8.GetBytes(new ReadOnlySpan<byte>(pInput + next, surrogatePair ? 2 : 1), encodedBytes);
encodedBytes = encodedBytes.Slice(encodedBytesCount);

foreach (byte b in encodedBytes)
{
    UriHelper.EscapeAsciiChar((char)b, ref dest);
}
```

Reply:
@MihaZupan 25% is a solid indicator and I think it's caused by this [ReadOnlySpan ctor](https://gist.github.com/MihaZupan/628bf37fe3d815183bb3d93fba337afc#file-unsafe-cs-L58) because I noticed a per degradation related to this while working on escaping/unescaping optimization.

Reply:
@scalablecory Yes, but the alternative to bound checks is an extra Slice, so that doesn't really change the numbers.

|    Method |     Mean |    Error |   StdDev |
|---------- |---------:|---------:|---------:|
|      Span | 41.10 us | 0.535 us | 0.500 us |
| SpanSlice | 41.78 us | 0.573 us | 0.508 us |
|    Unsafe | 33.18 us | 0.280 us | 0.248 us |

Reply:
Can you copy/paste the benchmark code? I'd like to poke at it myself a bit later.

Reply:
Oh, I see now. We're only encoding a single code point at a time here, so the overhead of *any* bounds check is going to be pretty big.

Reply:
I extracted just the needed files to https://github.com/MihaZupan/BenchmarkPR32025

Reply:
Oh, you're not actually measuring System.Private.Uri, but rather copying the source out into the benchmark?  That's not going to be equivalent.  For example, we explicitly clear the localsinit flag for all framework assemblies, but that won't happen for your code compiled into your benchmark, which means things like stackalloc are going to be more expensive.

Reply:
`Encoding.UTF8.GetBytes` is a bit heavyweight for this. I'd instead recommend a slight variation of what @scalablecory recommended:

```cs
Span<byte> encodedBytes = stackalloc byte[MaxNumberOfBytesEncoded];

Rune rune = (surrogatePair) ? new Rune(pInput[next], pInput[next + 1]) : new Rune(pInput[next]);
int encodedBytesCount = rune.EncodeToUtf8(encodedBytes);
encodedBytes = encodedBytes.Slice(0, encodedBytesCount);

foreach (byte b in encodedBytes)
{
    UriHelper.EscapeAsciiChar((char)b, ref dest);
}
```

Reply:
Removing localsinit reduces the gap substantially.
Using @GrabYourPitchforks 's approach beats all above 👍 

|    Method |     Mean |    Error |   StdDev |
|---------- |---------:|---------:|---------:|
|    Unsafe | 33.68 us | 0.401 us | 0.356 us |
|      Span | 35.34 us | 0.675 us | 0.853 us |
| SpanSlice | 35.23 us | 1.043 us | 0.871 us |
|      Rune | 28.13 us | 0.547 us | 0.512 us |

I'll make the change.

Reply:
To satisfy my curiosity, I added a benchmark with `Rune.EncodeToUtf8` encoding to a `byte*`, avoiding spans. It performs ~8% better than the Rune benchmark above.
(I am _not_ saying that I prefer it over the Rune & Span based one)